packages/devtools_app/lib/src/history_manager.dart

```diff
@@ -0,0 +1,78 @@
+// Copyright 2021. The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+
+class HistoryManager<T> {
+  final _current = ValueNotifier<T>(null);
+  final _history = <T>[];
+  int _historyIndex = -1;
+
+  /// Returns true if there is a previous historical item available on the
+  /// stack.
+  bool get hasPrevious {
+    return _history.isNotEmpty && _historyIndex > 0;
+  }
+
+  /// Returns true if there is a next historical item available on the stack.
+  bool get hasNext {
+    return _history.isNotEmpty && _historyIndex < _history.length - 1;
+  }
+
+  /// Return the next historical item on the stack.
+  ///
+  /// Throws [StateError] if there's no next item available.
+  T moveForward() {
```

Comment:
I don't think these methods need to return anything. Doesn't look like the return values are used anywhere, nor should they be really because we want `current` to be the source of data for parts of the code that care about the `_current.value`. Can we change the return to void for `moveForward`, `moveBack`, and `pop`?